### PR TITLE
Change the userspace proxy to wait for the node record

### PR DESCRIPTION
**Pull request by knobunc**
The userspaceproxy setup races with the node registration.  The userspace proxy setup tries to read the node record to retrieve the address for the node.  If the registration has not happened yet the read fails, but the code proceeds despite the failure.

The fix is to add a retry loop (with a backoff) to keep reading until the node record is present.

The bug is masked when a network plugin is used because it typically needs the node record to be present and it runs first, so by the time the proxy is set up, there is a node record.

Fixes bug 1519991 (https://bugzilla.redhat.com/show_bug.cgi?id=1519991)

@openshift/networking PTAL

### CODE REVIEW

File: pkg/cmd/server/kubernetes/network/network.go [hunk 1]
```diff
@@ -232,17 +236,28 @@ func (c *NetworkConfig) RunProxy() {
...
+	// This will wait 2, 4, 8, ... 256 seconds, for a total of 8.5 mins
+	nodeWaitBackoff := utilwait.Backoff{
+		Duration: 2 * time.Second,
+		Factor:   2,
+		Steps:    8,
 	}
-	nodeIP, err = utilnode.GetNodeHostIP(node)
+	utilwait.ExponentialBackoff(nodeWaitBackoff, func() (bool, error) {
+		node, nodeErr = client.Nodes().Get(hostname, metav1.GetOptions{})
+		return nodeErr == nil, nil
```
Comment (rajatchopra) [inline pkg/cmd/server/kubernetes/network/network.go]: Ideally, 'return nodeErr==nil, nodeErr'

File: pkg/cmd/server/kubernetes/network/network.go [hunk 2]
```diff
@@ -232,17 +236,28 @@ func (c *NetworkConfig) RunProxy() {
...
+	})
+	if nodeErr != nil {
+		return nil, fmt.Errorf("Failed to retrieve node info: %v", nodeErr)
+	}
+
+	nodeIP, err := utilnode.GetNodeHostIP(node)
```
Comment (rajatchopra) [inline pkg/cmd/server/kubernetes/network/network.go]: Do we actually fail to 'Get' the node itself, or do we fail to get its HostIP?

Comment (rajatchopra) [reply pkg/cmd/server/kubernetes/network/network.go]: I presume its only the 'Get' which is racing and we do not have to wait for HostIP fetching (status etc).

Comment (knobunc) [reply pkg/cmd/server/kubernetes/network/network.go]: Yeah, it's only the Get that is the problem.  We are reading the IP address that is reported at node creation so it's there when the node is there.

File: pkg/cmd/server/kubernetes/network/network.go (hunk 1 above)
Comment (knobunc) [reply pkg/cmd/server/kubernetes/network/network.go]: I can't... per vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go if I return an error, the backoff ends and the error is returned.

The first `nodeErr == nil` is the backoff termination condition, and we want that to be true when there are no errors.  I update the nodeErr defined outside the condition function so that we can get the error without messing with the backoff termination.

**Comment by knobunc** [comment]
@rajatchopra please re-review.

@smarterclayton is this reasonable?  It's blocking @stevekuznetsov.  Thanks.

File: pkg/cmd/server/kubernetes/network/network.go [hunk 3]
```diff
@@ -232,17 +236,30 @@ func (c *NetworkConfig) RunProxy() {
...
-	if err != nil {
-		glog.Warningf("Failed to retrieve node info: %v", err)
-		return nil
+func getNodeIP(client kv1core.CoreV1Interface, hostname string) (net.IP, error) {
+	var node *v1.Node
+	var nodeErr error
+
+	// We may beat the thread that causes the node object to be created,
+	// so if we can't get it, then we need to wait.
+	// This will wait 2, 4, 8, ... 256 seconds, for a total of 8.5 mins
```
Comment (pravisankar) [inline pkg/cmd/server/kubernetes/network/network.go]: For the first iteration it will not wait, so the times will be
0, 2, 4, 8, ... 128 secs, for a total of approx 4 mins

I found this issue in other places in SDN code and I plan to submit a fix in a separate PR.

File: pkg/cmd/server/kubernetes/network/network.go [hunk 4]
```diff
@@ -232,17 +236,30 @@ func (c *NetworkConfig) RunProxy() {
...
+		Factor:   2,
+		Steps:    8,
 	}
-	nodeIP, err = utilnode.GetNodeHostIP(node)
+	utilwait.ExponentialBackoff(nodeWaitBackoff, func() (bool, error) {
+		node, nodeErr = client.Nodes().Get(hostname, metav1.GetOptions{})
+		return nodeErr == nil, nil
+	})
+	if nodeErr != nil {
+		return nil, fmt.Errorf("Failed to retrieve node info: %v", nodeErr)
```
Comment (pravisankar) [inline pkg/cmd/server/kubernetes/network/network.go]: fmt.Errorf("failed ...) [error starts with lower case]

File: pkg/cmd/server/kubernetes/network/network.go [hunk 5]
```diff
@@ -92,7 +92,11 @@ func (c *NetworkConfig) RunProxy() {
 	case componentconfig.ProxyModeIPTables:
 		glog.V(0).Info("Using iptables Proxier.")
 		if bindAddr.Equal(net.IPv4zero) {
-			bindAddr = getNodeIP(c.ExternalKubeClientset.CoreV1(), hostname)
+			var err error
+			bindAddr, err = getNodeIP(c.ExternalKubeClientset.CoreV1(), hostname)
```
Comment (pravisankar) [inline pkg/cmd/server/kubernetes/network/network.go]: This change is in iptables proxy and not userspace proxy, description talks about userspace proxy issue? (this could be lack of my understanding)

File: pkg/cmd/server/kubernetes/network/network.go [hunk 6]
```diff
@@ -232,17 +236,30 @@ func (c *NetworkConfig) RunProxy() {
...
+	})
+	if nodeErr != nil {
+		return nil, fmt.Errorf("Failed to retrieve node info: %v", nodeErr)
+	}
+
+	nodeIP, err := utilnode.GetNodeHostIP(node)
 	if err != nil {
-		glog.Warningf("Failed to retrieve node IP: %v", err)
-		return nil
+		return nil, fmt.Errorf("Failed to retrieve node IP: %v", err)
```
Comment (pravisankar) [inline pkg/cmd/server/kubernetes/network/network.go]: fmt.Errorf("failed ...) [error starts with lower case]

File: pkg/cmd/server/kubernetes/network/network.go (hunk 5 above)
Comment (knobunc) [reply pkg/cmd/server/kubernetes/network/network.go]: No, that's a good point.  I am being imprecise.  What happens is that the iptables proxier needs to know the ip address, but it only matters when the userspace proxy is being used.  The iptables points to that ip when it sends traffic to the port for the userspace proxy.  So it doesn't show up in the testing as a problem until the unidler is used.

File: pkg/cmd/server/kubernetes/network/network.go (hunk 3 above)
Comment (knobunc) [reply pkg/cmd/server/kubernetes/network/network.go]: Ah, thanks!  I'll update.

File: pkg/cmd/server/kubernetes/network/network.go (hunk 1 above)
Comment (knobunc) [reply pkg/cmd/server/kubernetes/network/network.go]: Done, thanks

File: pkg/cmd/server/kubernetes/network/network.go [hunk 7]
```diff
@@ -232,17 +236,30 @@ func (c *NetworkConfig) RunProxy() {
...
+		Duration: 2 * time.Second,
+		Factor:   2,
+		Steps:    8,
 	}
-	nodeIP, err = utilnode.GetNodeHostIP(node)
+	utilwait.ExponentialBackoff(nodeWaitBackoff, func() (bool, error) {
+		node, nodeErr = client.Nodes().Get(hostname, metav1.GetOptions{})
+		return nodeErr == nil, nil
+	})
+	if nodeErr != nil {
```
[the remark this answers is not in the document]
Comment (knobunc) [reply pkg/cmd/server/kubernetes/network/network.go]: But, don't I want to return `nodeErr` rather than the timeout error?  I think the error really is that we couldn't get the node object.  Returning `ErrWaitTimeout` just obscures the problem.

File: pkg/cmd/server/kubernetes/network/network.go (hunk 5 above)
Comment (pravisankar) [reply pkg/cmd/server/kubernetes/network/network.go]: Okay understood, thanks for the explanation.

File: pkg/cmd/server/kubernetes/network/network.go [hunk 8]
```diff
@@ -232,17 +237,37 @@ func (c *NetworkConfig) RunProxy() {
...
-	if err != nil {
-		glog.Warningf("Failed to retrieve node info: %v", err)
-		return nil
+func getNodeIP(client kv1core.CoreV1Interface, hostname string) (net.IP, error) {
+	var node *v1.Node
+	var nodeErr error
+
+	// We may beat the thread that causes the node object to be created,
+	// so if we can't get it, then we need to wait.
+	// This will wait 0, 2, 4, 8, ... 256 seconds, for a total of 8.5 mins
```
Comment (pravisankar) [inline pkg/cmd/server/kubernetes/network/network.go]: Do we really want to wait for ~8 mins (seems a bit long)?
In openshift SDN node, we wait for local subnet which requires node registration and master reacting with this event and creating hostSubnet. Total wait timeout in this case is only 1 min (actually 30 secs because of the bug), this may be low. We did not see issues with this, to be on safer side (when node loaded) 2 to 3 mins may suffice.

Comment (knobunc) [reply pkg/cmd/server/kubernetes/network/network.go]: My testing showed registration happened by the 2nd or 3rd attempt... so 2-4 seconds.  I'll lower it.

**Comment by knobunc** [comment]
@stevekuznetsov -- Any idea why a test that doesn't show up is blocking this?  Thanks.

**Comment by stevekuznetsov** [comment]
@knobunc unclear why it's hung on that but in any case you need LGTM and approval for this

**Comment by stevekuznetsov** [comment]
Amazing!!! Let's get this into `ose@master` and backported to `ose@enterprise-3.7` and `ose@enterprise-3.8` and `ose@enterprise-3.9`
